### DISCUSSION TITLE
Sound way of determining default i2c bus

### DISCUSSION THREAD

**Discussion by sgparry** [mattjlewis/diozero#145] [Q&A]
Hi,
As I understand it, most Pis use bus 1 for i2c by default, but some models use 0. What is the best way, using diozero, to determine which?
Thanks

**Comment by sgparry** [comment]
Not sure what planet I was on when I wrote this question - I meant i2c not DMA...

**Comment by EAGrahamJr** [comment]
Have you looked at https://www.diozero.com/utilityapps.html? The code for `com.diozero.sampleapps.I2CDetect` might get you somewhere -- at the minimum, you'd be able to scan a given controller to see if it finds any devices. :smiley:

**Comment by sgparry** [comment]
@EAGrahamJr thanks for the reply...
I need something lightweight; I am converting some Python library code that sets the default i2c bus for its constructors based on a version field (it is not clear as to which from the code). 99 times out of 100, it will be bus 1. Having googled around a bit, I think the following is sufficient in diozero:
```
    private static int getDefaultI2cBus() {
        BoardInfo boardInfo = DeviceFactoryHelper.getNativeDeviceFactory().getBoardInfo();
        String longName = boardInfo.getLongName();
        if (longName.equals("Raspberry Pi Model B v1.0") || longName.equals("Raspberry Pi Model B v1.1")) {
            return 0;
        } else {
            return 1;
        }
    }
```
The version strings correspond to legacy revisions 0002 and 0003. Can anyone confirm that this looks right or suggest better please?

**Comment by EAGrahamJr** [comment]
Well, there is the notion of not bothering to support such old boards :smiley:

If you're going that route, I wouldn't use hard-coded strings, but rather use the board's `pcbRevision` and select on the _constants_ `PCB_REV_1_0` and `PCB_REV_1_1` - that way your code shouldn't break if Matt decides to tweak the board info.

You could also construct an `I2CDevice` with the target device address and just see if it's there with a `probe`": "if it's not on bus 1, try bus 0".

**Comment by sgparry** [comment]
Hey, I'm not using them, but I think that 2 of the original Model A s I have may be V1.1!
I went for getLongName to avoid having to compare three pieces of information (who knows what values a clone might return) but I think I will split the comparison up.
If I were writing my code from scratch I would probably probe, but I want to keep behaviour as close to the original python code as possible. Thanks!

**Comment by sgparry** [comment]
Slight problem, all the constants and the getPcbRevision are only accessible on the internal API version of the interface.

**Reply by EAGrahamJr** [reply]
:thinking:

**Comment by mattjlewis** [comment]
You should be able to cast the BoardInfo returned from the device factory getBoardInfo method to PiBoardInfo and then call the public operation getPcbRevision. The PCB_REV_ constants are all public in RaspberryPiBoardInfoProvider
